app/src/main/java/com/hedvig/app/feature/insurance/ui/tab/InsuranceFragment.kt

```
@@ -0,0 +1,80 @@
+import com.hedvig.app.feature.loggedin.ui.ScrollPositionListener
+import com.hedvig.app.feature.settings.MarketManager
+import com.hedvig.app.util.extensions.viewLifecycle
+import com.hedvig.app.util.extensions.viewLifecycleScope
+import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import org.koin.android.ext.android.inject
+import org.koin.androidx.viewmodel.ext.android.sharedViewModel
+
+class InsuranceFragment : Fragment(R.layout.fragment_insurance) {
+    private val insuranceViewModel: InsuranceViewModel by sharedViewModel()
+    private val loggedInViewModel: LoggedInViewModel by sharedViewModel()
+    private val tracker: InsuranceTracker by inject()
+    private val marketManager: MarketManager by inject()
+    private val binding by viewBinding(FragmentInsuranceBinding::bind)
+    private var scroll = 0
+
+    override fun onResume() {
+        super.onResume()
+        loggedInViewModel.onScroll(scroll)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        binding.insuranceRecycler.apply {
+            scroll = 0
+            addOnScrollListener(
+                ScrollPositionListener(
+                    { scrollPosition ->
+                        scroll = scrollPosition
+                        loggedInViewModel.onScroll(scrollPosition)
+                    },
+                    viewLifecycleOwner
+                )
+            )
+            adapter = InsuranceAdapter(tracker, marketManager, insuranceViewModel::load)
+        }
+        insuranceViewModel
+            .data
+            .flowWithLifecycle(viewLifecycle)
+            .onEach { bind(it) }
+            .launchIn(viewLifecycleScope)
+    }
+
+    private fun bind(viewState: InsuranceViewModel.ViewState) {
+        val adapter = binding.insuranceRecycler.adapter as? InsuranceAdapter ?: return
```

Should this throw instead of returning?

---

This code only looks new, it's actually exactly the code that was in `InsuranceFragment` before the file was moved 😬

---

Aha!